plotly/tools.py

```diff
@@ -1451,6 +1451,697 @@ class FigureFactory(object):
     """
 
     @staticmethod
+    def _calc_stats(data):
+        """
+        Calculate statistics for use in violin plot.
+        """
+        import numpy as np
+
+        x = np.asarray(data, np.float)
+        vals_min = np.min(x)
+        vals_max = np.max(x)
+        q2 = np.percentile(x, 50, interpolation='linear')
+        q1 = np.percentile(x, 25, interpolation='lower')
+        q3 = np.percentile(x, 75, interpolation='higher')
+        IQR = q3 - q1
+        whisker_dist = 1.5 * IQR
+
+        # in order to prevent drawing whiskers outside the interval
+        # of data one defines the whisker positions as:
+        d1 = np.min(x[x >= (q1 - whisker_dist)])
+        d2 = np.max(x[x <= (q3 + whisker_dist)])
+        return vals_min, vals_max, q1, q2, q3, d1, d2
```

Comment:
🐄 Can we return a `dict` from here? I could see us wanting to return more things from `stats` in the future and it's not obvious that simply appending these stats will make sense, organizationally.

```
return {
    'min': np.min(x),
    'max': np.max(x),
    ...
}
```